core/src/main/kotlin/org/evomaster/core/problem/external/service/ExternalServices.kt

```diff
@@ -1,26 +1,169 @@
+     * while checking the availability. If not available will
+     * generate a new one.
+     */
+    private fun generateRandomAvailableAddress() : String {
+        val ip = generateRandomIPAddress()
+        if (isAddressAvailable(ip, WIREMOCK_PORT)) {
+            return ip
+        }
+        return generateRandomAvailableAddress()
+    }
+
     fun getExternalServices() : List<ExternalServiceInfo> {
         return externalServiceInfos
     }
+
+    /**
+     * Default IP address will be a randomly generated IP
+     */
+    private fun getIP() : String {
+        val ip: String
+        when (config.externalServiceIPSelectionStrategy) {
+            // Although the default address will be a random, this
+            // option allows selecting explicitly
+            EMConfig.ExternalServiceIPSelectionStrategy.RANDOM -> {
+                ip = if (externalServiceInfos.size > 0) {
+                    getNextAvailableAddress()
+                } else {
+                    generateRandomAvailableAddress()
+                }
+            }
+            EMConfig.ExternalServiceIPSelectionStrategy.USER -> {
+                ip = if (externalServiceInfos.size > 0) {
+                    getNextAvailableAddress()
+                } else {
+                    config.externalServiceIP
+                }
+            }
+            else -> {
+                ip = if (externalServiceInfos.size > 0) {
+                    getNextAvailableAddress()
+                } else {
+                    generateRandomAvailableAddress()
+                }
+            }
+        }
+        return ip
+    }
+    /**
+     * Will initialise WireMock instance on a given IP address for a given port.
+     */
+    private fun initWireMockServer(address: String): WireMockServer {
+        val wm = WireMockServer(
+            WireMockConfiguration()
+                .bindAddress(address)
+                .port(WIREMOCK_PORT)
+                .extensions(ResponseTemplateTransformer(false)))
+        wm.start()
+
+        // to prevent from the 404 when no matching stub below stub is added
+        wm.stubFor(get(urlMatching("/.*"))
+            .atPriority(2)
+            .willReturn(aResponse()
+                .withStatus(200)
+                .withBody("Not found!!")))
+
+        return wm
+    }
+
+    /**
+     * Update the DNS cache with a different IP address for a given host to enable spoofing.
+     * If there is an entry already it'll skip from adding to the cache.
+     */
+    private fun updateDNSCache(host : String, address : String) {
+        val entry = DnsCacheManipulator.getDnsCache(host)
+        if (entry != null) {
+            DnsCacheManipulator.setDnsCache(host, address)
+        }
+    }
+
+    /**
+     * Clean up the DNS cache
+     */
+    fun cleanDNSCache() {
+        DnsCacheManipulator.clearDnsCache()
+    }
+
+    companion object {
+        private const val WIREMOCK_PORT : Int = 8080
```

Comment:
why this? shouldn't be based on the port called from the SUT?

Reply:
Most cases SUT port will be 80 or 443. To bind on those ports we need root privileges. Let me check this again, if it's working, I'll make the changes to use SUT port.